File: DP/targetSum.py
class Solution:
    def countSubsetSum(self, arr, n, s):
        t=[[1 if i==j==0 else 0 for i in range(s+1)] for j in range(n+1)]
        for i in range(1,n+1):
            for j in range(0, s+1):
                if arr[i-1]>j:
                    t[i][j]=t[i-1][j]
                else:
                    t[i][j]= t[i-1][j-arr[i-1]] + t[i-1][j]
        return t[n][s]
                
    
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        r=sum(nums)
        if (r+target)%2 !=0:
            return 0
        s1=(r+target)//2
        return self.countSubsetSum(nums, len(nums), abs(s1))

File: DP-CPP/targetSum.cpp
int mod = (int)1e9 + 7;
class Solution
{
public:
    int findTargetSumWays(vector<int> &arr, int d)
    {
        // Write your code here.
        int n = arr.size();

        int sum = 0;
        for (auto &it : arr)
        {
            sum += it;
        }
        if (sum < d || (sum - d) % 2)
        {
            return 0;
        }
        //     cout<<sum<<endl;

        int K = (sum - d) / 2;

        vector<vector<int>> dp(n, vector<int>(K + 1, 0));

        if (arr[0] == 0)
            dp[0][0] = 2;
        else
            dp[0][0] = 1;

        if (arr[0] != 0 && arr[0] <= K)
            dp[0][arr[0]] = 1;

        for (int ind = 1; ind < n; ind++)
        {
            for (int target = 0; target <= K; target++)
            {
                int nottake = dp[ind - 1][target];
                int take = 0;
                if (arr[ind] <= target)
                {
                    take += dp[ind - 1][target - arr[ind]];
                }
                dp[ind][target] = (take + nottake) % mod;
            }
        }
        return dp[n - 1][K];
    }
};